### PR TITLE
Widescreen mode based on Crispy Doom

### PR DESCRIPTION
**Pull request by rfomin**
This is incomplete, but it mostly works:

scrn1

What do you think? Should I continue?

Related issue: #94

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Impressive, thank you very much! In how far is this incomplete? Yes, of course you should continue. 😉

**Comment by rfomin** [comment]
> In how far is this incomplete?

It needs the support of wide assets. I haven't tested bunny scroll end ending screens, nothing major.

**Comment by fabiangreffrath** [comment]
I see. Would you like me to review this, so we can get it merged first and add wide asset support afterwards?

**Comment by rfomin** [comment]
Not now, I need to fix few things. I will tell you when I'm ready for review.

**Comment by fabiangreffrath** [comment]
Looking forward to it, please keep up the good work. 👍

**Comment by rfomin** [comment]
Ready to review.

**Comment by rfomin** [comment]
Will you include wide assets in Woof distribution? I propose this file structure:
```
wideassets/
├── wide_d1.wad
├── wide_d2.wad
├── wide_tnt.wad
└── wide_pl.wad
```

**Comment by fabiangreffrath** [comment]
> Will you include wide assets in Woof distribution? I propose this file structure:
> 
> ```
> wideassets/
> ├── wide_d1.wad
> ├── wide_d2.wad
> ├── wide_tnt.wad
> └── wide_pl.wad
> ```

Absolutely no. The assets are copyrighted by id software and they didn't grant me any license to distribute them.

**Comment by rfomin** [comment]
https://bethesda.net/en/game/doom-widescreen-mods
Also, GZDoom did it.

**Comment by fabiangreffrath** [comment]
I didn't even know this, thank you! Anyway, this is a free software project, licensed under the GPL. And I have taken measures to make sure that everything in this project complies to this license, e.g. by removing some of the embedded lumps that resembled parts of Doom's original artwork or by replacing the embedded dog sprites and sounds from Wolf-3D and the plasma ball sprites from the Doom Beta by strictly free alternatives. Now I am not going to dilute this achievement by adding non-free [1] data files that everybody could download himself anyway. I hope you understand.

[1] https://wiki.debian.org/FreeSoftware

**Comment by rfomin** [comment]
Okay, then I consider this complete. Except for multi-monitor support, because now we need screen sizes.

File: Source/i_video.c [hunk 1]
```diff
@@ -974,6 +979,62 @@ int cfg_aspectratio; // haleyjd 05/11/09: aspect ratio correction
...
+      //         w = 21;
+      //         h = 9;
+      //         break;
+      //     default:
+      //         break;
+      // }
+
+      SCREENWIDTH = w * ah / h;
+      // [crispy] make sure SCREENWIDTH is an integer multiple of 4 ...
+      SCREENWIDTH = (SCREENWIDTH + 3) & (int)~3;
```
Comment (fabiangreffrath) [inline Source/i_video.c:1029]: I have recently changed this to `SCREENWIDTH = (SCREENWIDTH + (crispy->hires ? 0 : 3)) & (int)~3;` in Crispy, please do the same here.

File: Source/f_finale.c [hunk 2]
```diff
@@ -685,21 +687,52 @@ void F_BunnyScroll (void)
       scrolled = 320;
   if (scrolled < 0)
       scrolled = 0;
-              
-  for ( x=0 ; x<SCREENWIDTH ; x++)
+
```
Comment (fabiangreffrath) [inline Source/f_finale.c:690]: One blank line too much. ;)

File: Source/m_menu.c [hunk 3]
```diff
@@ -1784,25 +1784,39 @@ void M_DrawBackground(char* patchname, byte *back_dest)
...
+	//     do
+	//       back_dest[i*2] = back_dest[i*2+SCREENWIDTH*2] =
+	// 	back_dest[i*2+1] = back_dest[i*2+SCREENWIDTH*2+1] = src[i];
+	//     while (--i>=0);
+	//     back_dest += 128;
+	//   }
+    for (int y = 0; y < SCREENHEIGHT<<1; y++)
+      for (int x = 0; x < SCREENWIDTH<<1; x += 2)
+      {
+          byte  dot = src[(((y>>1)&63)<<6) + ((x>>1)&63)];
```
Comment (fabiangreffrath) [inline Source/m_menu.c:1801]: `const byte` and one less space char, please.

File: Source/m_menu.c [hunk 4]
```diff
@@ -1784,25 +1784,39 @@ void M_DrawBackground(char* patchname, byte *back_dest)
...
-	for (x = 0 ; x < SCREENWIDTH/64 ; x++)
-	  {
-	    int i = 63;
-	    do
-	      back_dest[i*2] = back_dest[i*2+SCREENWIDTH*2] =
-		back_dest[i*2+1] = back_dest[i*2+SCREENWIDTH*2+1] = src[i];
-	    while (--i>=0);
-	    back_dest += 128;
-	  }
+ //      for (y = 0 ; y < SCREENHEIGHT ; src = ((++y & 63)<<6) + back_src,
```
Comment (fabiangreffrath) [inline Source/m_menu.c:1787]: Please comment out entire code ranges by using `/* ... */`, applies below as well.

File: Source/r_draw.c [hunk 5]
```diff
@@ -57,7 +57,7 @@ int  viewwindowx;
 int  viewwindowy; 
 byte *ylookup[MAXHEIGHT]; 
 int  columnofs[MAXWIDTH]; 
-int  linesize = SCREENWIDTH;  // killough 11/98
+int  linesize;// = SCREENWIDTH;  // killough 11/98
```
Comment (fabiangreffrath) [inline Source/r_draw.c:60]: Let's please still initialize this with `ORIGWIDTH` here.

File: Source/r_main.c [hunk 6]
```diff
@@ -238,6 +239,12 @@ fixed_t R_ScaleFromGlobalAngle(angle_t visangle)
     64*FRACUNIT : num < 256 ? 256 : num : 64*FRACUNIT;
 }
 
+
```
Comment (fabiangreffrath) [inline Source/r_main.c:242]: Excess blank line.

File: Source/st_stuff.c [hunk 7]
```diff
@@ -323,17 +323,85 @@ extern char     *mapnames[];
...
+        byte *src;
+        byte *dest;
+        const char *name = (gamemode == commercial) ? "GRNROCK" : "FLOOR7_2";
+
+        src = W_CacheLumpNum(firstflat + R_FlatNumForName(name), PU_CACHE);
+        dest = screens[BG];
+
+        if (hires)
+        {
+          for (int y = (SCREENHEIGHT-ST_HEIGHT)<<1; y < SCREENHEIGHT<<1; y++)
```
Comment (fabiangreffrath) [inline Source/st_stuff.c:345]: `x` and `y` are already declared above.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Very good, mostly style issues left.

**Comment by fabiangreffrath** [comment]
Thank you very much! Wow, this is really exciting...

**Pull request merged**